TombEngine/Game/control/box.cpp

```diff
@@ -1432,69 +1438,101 @@ void FindAITarget(CreatureInfo* creature, short objectNumber)
+		return false;
 
-	AI_OBJECT* foundObject = nullptr;
+	auto& creature = *GetCreatureInfo(&item);
+
+	const AI_OBJECT* foundAIObject = nullptr;
 
-	for (auto& aiObject : g_Level.AIObjects)
+	for (const auto& aiObject : g_Level.AIObjects)
 	{
-		if (aiObject.objectNumber == objectNumber &&
-			aiObject.triggerFlags == ocb &&
-			aiObject.roomNumber != NO_VALUE)
+		// Check if object IDs match.
+		if (aiObject.objectNumber != data.ObjectID)
+			continue;
+
+		// Check if room is valid.
+		if (aiObject.roomNumber == NO_VALUE)
+			continue;
+
+		// Check if distance is valid.
+		if (data.CheckDistance)
+		{
+			if (Vector3i::Distance(item.Pose.Position, aiObject.pos.Position) > data.DistanceMax)
+				continue;
+		}
+
+		// Check if OCBs match (useful for pathfinding).
+		if (data.CheckOcb)
 		{
-			int* zone = g_Level.Zones[(int)creature->LOT.Zone][(int)FlipStatus].data();
+			if (aiObject.triggerFlags != data.Ocb)
+				continue;
+		}
+
+		// Check if zone IDs match.
+		if (data.CheckSameZone)
+		{
+			int* zone = g_Level.Zones[(int)creature.LOT.Zone][(int)FlipStatus].data();
 			auto* room = &g_Level.Rooms[item.RoomNumber];
 
-			item.BoxNumber = GetSector(room, item.Pose.Position.x - room->Position.x, item.Pose.Position.z - room->Position.z)->PathfindingBoxID;
+			// NOTE: Avoid changing box ID of item or AI item so a local variable isn't required when searching for AI object near it.
+			int boxID = GetSector(room, item.Pose.Position.x - room->Position.x, item.Pose.Position.z - room->Position.z)->PathfindingBoxID;
 			room = &g_Level.Rooms[aiObject.roomNumber];
-			aiObject.boxNumber = GetSector(room, aiObject.pos.Position.x - room->Position.x, aiObject.pos.Position.z - room->Position.z)->PathfindingBoxID;
-
-			if (item.BoxNumber == NO_VALUE || aiObject.boxNumber == NO_VALUE)
-				return;
+			int aiBoxID = GetSector(room, aiObject.pos.Position.x - room->Position.x, aiObject.pos.Position.z - room->Position.z)->PathfindingBoxID;
 
-			if (checkSameZone && (zone[item.BoxNumber] != zone[aiObject.boxNumber]))
-				return;
+			// Box is invalid or zones don't match; continue.
+			if (boxID == NO_VALUE || aiBoxID == NO_VALUE)
+				continue;
 
-			// Don't check for same zone. Needed for Sophia Leigh.
-			foundObject = &aiObject;
+			// Zone is invalid; continue.
+			if (zone[boxID] != zone[aiBoxID])
+				continue;
 		}
-	}
-
-	if (foundObject == nullptr)
-		return;
 
-	auto& aiItem = *creature->AITarget;
+		// HACK: Don't check for matching zone. Needed for Sophia Leigh.
+		foundAIObject = &aiObject;
+	}
 
-	creature->Enemy = &aiItem;
+	if (foundAIObject == nullptr)
+		return false;
 
-	aiItem.ObjectNumber = foundObject->objectNumber;
-	aiItem.RoomNumber = foundObject->roomNumber;
-	aiItem.Pose.Position = foundObject->pos.Position;
-	aiItem.Pose.Orientation.y = foundObject->pos.Orientation.y;
-	aiItem.Flags = foundObject->flags;
-	aiItem.TriggerFlags = foundObject->triggerFlags;
-	aiItem.BoxNumber = foundObject->boxNumber;
+	auto aiItem = ItemInfo{};
```

Comment:
Why new `ItemInfo` is created and returned by this function?

Reply:
It's required if i want the enemy to move toward it, creature info expect a ItemInfo* in FindAITargetObject()